src/Microsoft.Extensions.Primitives/StringSegment.cs

```diff
@@ -0,0 +1,162 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.Extensions.Primitives
+{
+    public struct StringSegment : IEquatable<StringSegment>
+    {
+        public StringSegment(string buffer, int offset, int count)
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException(nameof(buffer));
+            }
+
+            if (offset < 0 || offset >= buffer.Length)
+            {
+                throw new IndexOutOfRangeException(nameof(offset));
+            }
+
+            if (count < 0 || offset + count >= buffer.Length)
+            {
+                throw new IndexOutOfRangeException(nameof(count));
+            }
+
+            Buffer = buffer;
+            Offset = offset;
+            Count = count;
+        }
+
+        public string Buffer { get; }
+
+        public int Offset { get; }
+
+        public int Count { get; }
+
+        public string Value
+        {
+            get { return HasValue ? Buffer.Substring(Offset, Count) : null; }
+        }
+
+        public bool HasValue
+        {
+            get { return Buffer != null; }
+        }
+
+        public bool Equals(StringSegment other)
```

Comment:
Change Equals and GetHashCode to only look at the in-range data. Also add an overload of Equals(StringSegment, StringComparison)